snekbox/nsjail.py

```diff
@@ -152,8 +166,28 @@ def python3(
                 *nsjail_args,
             )
 
-        with NamedTemporaryFile() as nsj_log:
-            args = (
+        with NamedTemporaryFile() as nsj_log, MemFS(self.memfs_instance_size) as fs:
+            # Add the temp dir to be mounted as cwd
+            nsjail_args = (
+                # Mount a tmpfs at /dev/shm to support multiprocessing
+                "--mount",
+                # src:dst:fs_type:options
+                f"{fs.shm}:/dev/shm:tmpfs:size={fs.instance_size}",
```

Comment:
I'm struggling to work out exactly what this does, does this create a new tmpfs (I would guess it does?)? If it does, then surely the `src` does nothing? I think the approach in #113 of defining `/dev/shm` in `snekbox.cfg` and just letting nsjail handle it makes more sense.

Reply:
My understanding (from skimming over the barely existant documentation and confusing cpp parser source, is that this mounts a regular folder (in this case backed by tmpfs on the docker side) onto nsjail as a `tmpfs`, so it isn't actually a tmpfs created on the nsjail side, but a fake one? This might be completely wrong, I'm not sure.

However the space here is usable to the full limit of our memfs allocation, and not the memory limit of the instance, so that would suggest it's not using nsjail memory? 

In any case there seems to be 0 difference in whether this is a bindmount (like `home`) or `tmpfs` mount as far as I can tell? I opted for `tmpfs` as it seems more "realistic" as far as the nsjail/python side is concerned on what a `/dev/shm` should be.

Reply:
> My understanding (from skimming over the barely existant documentation and confusing cpp parser source), is that this mounts a regular folder (in this case backed by tmpfs on the docker side) onto nsjail as a tmpfs, so it isn't actually a tmpfs created on the nsjail side, but a fake one? This might be completely wrong, I'm not sure.

I did some testing and I think my guess was correct and it just creates a new tmpfs ignoring the source. To demonstrate this you can modify the code to this:
```python
            (fs.shm / "test.txt").write_text("hello world")
            log.info(list(fs.shm.iterdir()))

            try:
                nsjail = subprocess.Popen(
                    args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True
                )
            except ValueError:
                return EvalResult(args, None, "ValueError: embedded null byte")
            
            log.info(list(fs.shm.iterdir()))
```
And try sending this to snekbox:
```python
import os
open('/dev/shm/thinggggggg.txt', 'w').write('hello world')
print(os.listdir('/dev/shm'))
```
In the logs you'll get something like
```
snekbox_dev  | 2022-11-18 23:23:14,519 |     8 |                 snekbox.nsjail |     INFO | [PosixPath('/memfs/0562f335-93b6-4db6-89d0-1ff580b58921/dev/shm/test.txt')]  
snekbox_dev  | 2022-11-18 23:23:14,520 |     8 |                 snekbox.nsjail |     INFO | [PosixPath('/memfs/0562f335-93b6-4db6-89d0-1ff580b58921/dev/shm/test.txt')] 
```
and in the stdout response from snekbox you'll get
```
['thinggggggg.txt']
```
I also tried replacing `{fs.shm}` with random characters in `f"{fs.shm}:/dev/shm:tmpfs:size={fs.instance_size}"` and it didn't seem to affect how it worked.

> In any case there seems to be 0 difference in whether this is a bindmount (like home) or tmpfs mount as far as I can tell? I opted for tmpfs as it seems more "realistic" as far as the nsjail/python side is concerned on what a /dev/shm should be.

There is a difference (bindmount uses the existing tmpfs), and whilst it doesn't really matter in this case, this current code is confusing (and the code to create `fs.shm` is redundant) so I do think the approach in #113 is better.

Reply:
Okay cool thanks for the testing and info. Yeah I agree in that case the config approach in #113 is probably better then and will be remove the shm mounting in this PR.

Reply:
Removed in [9327283](https://github.com/python-discord/snekbox/pull/159/commits/932728308f6773bb525028281ccf1a598d20786b)